ADBDEV-5473 Truncate temporary relation files before unlink

dnskvlnk opened the pull request:
Truncate temporary relation files before unlink

Problem description:
After the drop of a temporary heap relation, disk space was not released to the
OS for several minutes. The expected behavior is immediate disk space release.

Root cause:
Background writer process kept file descriptors of the relation open until the
next checkpoint. The next checkpoint occurred after a timeout, which could be up
to CheckPointTimeout seconds (300 by default). Only temporary relations suffered
from it, because in mdunlinkfork segment files were truncated for all
non-temporary relations.

Fix:
The mdunlinkfork function was changed to call do_truncate on all segment files
for all relations, including temporary.

Modified due to conflicts:
src/test/regress/input/tablespace.source
src/test/regress/output/tablespace.source

(cherry picked from commit https://github.com/arenadata/gpdb/commit/d622b980ce2b0eb3c2fa63b8c4f1ec29d7c01930)

silent-observer commented on src/test/regress/output/tablespace.source:
```diff
@@ -737,3 +737,78 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
 DROP TABLE tablespace_part;
 DROP TABLESPACE myts;
+DROP ROLE tablespace_testuser1;
+DROP ROLE tablespace_testuser2;
```
This is not present in the input

dnskvlnk replied:
Fixed

silent-observer commented on src/test/regress/input/tablespace.source:
```diff
@@ -333,3 +333,28 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
+SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+
+-- Checking that the temp heap table doesn't leave non-zero files.
+CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
```
The test fails because testspace doesn't exist

dnskvlnk replied:
Fixed

silent-observer reviewed (commented):
This should be cherry-picked with -x to preserve previous commit hash in the message, or you can add it manually

Commit:
Truncate temporary relation files before unlink

Problem description:
After the drop of a temporary heap relation, disk space was not released to the
OS for several minutes. The expected behavior is immediate disk space release.

Root cause:
Background writer process kept file descriptors of the relation open until the
next checkpoint. The next checkpoint occurred after a timeout, which could be up
to CheckPointTimeout seconds (300 by default). Only temporary relations suffered
from it, because in mdunlinkfork segment files were truncated for all
non-temporary relations.

Fix:
The mdunlinkfork function was changed to call do_truncate on all segment files
for all relations, including temporary.

Modified due to conflicts:
src/test/regress/input/tablespace.source
src/test/regress/output/tablespace.source

(cherry picked from commit d622b980ce2b0eb3c2fa63b8c4f1ec29d7c01930)

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-5473

Commit:
Fix path in test output

Commit:
Fix tablespace test by resetting default_tablespace

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-5473

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-5473

whitehawk commented:
The authorship of the original commit is lost.

silent-observer commented on src/test/regress/output/tablespace.source:
```diff
@@ -737,3 +737,63 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
+ Success:
+(3 rows)
+
+CREATE TABLESPACE testtablespace LOCATION '/home/gpadmin/gpdb_src/src/test/regress/testtablespace';
```
Please use `@testtablespace@` here too
